File: sdks/python/src/echo_agent_sdk/sandbox_resource_values.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ResourceLimits:
    """Sandbox resource policy value; process execution remains Rust-owned."""

    cpu_time_secs: int | None
    memory_bytes: int | None
    max_output_bytes: int | None
    max_processes: int | None
    network: bool
    read_only_paths: tuple[str, ...] = ()
    writable_paths: tuple[str, ...] = ()

    @classmethod
    def default(cls) -> ResourceLimits:
        return cls(30, 256 * 1024 * 1024, 1024 * 1024, 64, False)

    @classmethod
    def strict(cls) -> ResourceLimits:
        return cls(10, 64 * 1024 * 1024, 256 * 1024, 8, False)

    @classmethod
    def unrestricted(cls) -> ResourceLimits:
        return cls(None, None, None, None, True)

File: sdks/typescript/src/sandbox_resource_values.ts
export type ResourceLimitsValue = Readonly<{
  cpuTimeSecs?: bigint;
  memoryBytes?: bigint;
  maxOutputBytes?: bigint;
  maxProcesses?: number;
  network: boolean;
  readOnlyPaths: readonly string[];
  writablePaths: readonly string[];
}>;

function freeze(value: ResourceLimitsValue): ResourceLimitsValue {
  return Object.freeze({
    ...value,
    readOnlyPaths: Object.freeze([...value.readOnlyPaths]),
    writablePaths: Object.freeze([...value.writablePaths]),
  });
}

/** Sandbox resource policy value; process execution remains Rust-owned. */
export const ResourceLimits = Object.freeze({
  default(): ResourceLimitsValue {
    return freeze({ cpuTimeSecs: 30n, memoryBytes: 256n * 1024n * 1024n, maxOutputBytes: 1n * 1024n * 1024n, maxProcesses: 64, network: false, readOnlyPaths: [], writablePaths: [] });
  },
  strict(): ResourceLimitsValue {
    return freeze({ cpuTimeSecs: 10n, memoryBytes: 64n * 1024n * 1024n, maxOutputBytes: 256n * 1024n, maxProcesses: 8, network: false, readOnlyPaths: [], writablePaths: [] });
  },
  unrestricted(): ResourceLimitsValue {
    return freeze({ network: true, readOnlyPaths: [], writablePaths: [] });
  },
});
